### DISCUSSION TITLE
How do I upload an object as a json file from node.js?

### DISCUSSION THREAD

**Discussion by 72L** [supabase/supabase#2446] [Questions] [answered]
Could anyone give an example of uploading an object to Supabase storage as a JSON file?

I'd like to upload an object from a backend node.js server. Therefore, I cannot use the client-side `File` interface.

The objects are small enough to fit in memory, so ideally this process would stay all in memory so I don't have to write a file to disk.

Thanks!

**Comment by burggraf** [comment]
May I ask why you are trying to upload `JSON` objects to storage instead of storing them in PostgreSQL?  If they objects are small, PostgreSQL would handle them quite easily.  Just use a `JSONB` column type for this.

**Reply by 72L** [reply]
I'd be interested to know how to do this regardless of my current business use case, as I'm sure there are situations where storage would make sense. 

For example, if I want to save cost, Storage seems to be lower cost since I can store more files.

**Reply by burggraf** [reply]
Storage is generally designed for binary files, but I supposed you could create the `JSON` as a text file (`filename.json`) and upload that text file.

Storing JSON in a database is going to be cheaper than using Storage, though, because every item in storage also has a related database record as well which is used for security purposes (setting permissions, etc.)  But I wouldn't worry about pricing on storing small JSON records.  There's no limit to the number of records in your database, pricing only goes by size, and if the JSON is relatively small, even on the free tier you'd have tons and tons of room in your database.  :)

**Comment by thebengeu** [comment, marked as answer]
Since supabase/storage-js#5 was resolved, implementing server side storage upload, you can now use a Node.js `Buffer` or `Stream` (and many other types) with [from.upload()](https://supabase.io/docs/reference/javascript/storage-from-upload):

```
await supabase
  .storage
  .from('bucket')
  .upload('folder/file.json', jsonFile, {contentType: 'application/json'})
```

**Reply by 72L** [reply]
so would it be:

```
await supabase
  .storage
  .from('bucket')
  .upload(
    'folder/file.json', 
    Buffer.from(JSON.stringify({key: "value"})), 
    {contentType: 'application/json'}
  )
```

**Reply by thebengeu** [reply]
Yes, I've verified that works. A plain string would work fine too:

```
await supabase
  .storage
  .from('bucket')
  .upload(
    'folder/file.json', 
    JSON.stringify({key: "value"}), 
    {contentType: 'application/json'}
  )
```

However, due to supabase/storage-api#43, files uploaded with content type `application/json` and `text/plain` will be 0 bytes, for existing projects. If you create a new project, the uploads should work fine. The fix will be deployed for existing projects probably by the end of this week.

**Reply by 72L** [reply]
I'm trying it now, and it doesn't work. 

I get a response:
```
{"data":null,"error":{}}
```

And when I go to the dashboard at https://app.supabase.io/ there is no file in the bucket 

Using supabase here https://www.npmjs.com/package/supabase version 0.5.0

**Reply by thebengeu** [reply]
Do you mean https://www.npmjs.com/package/@supabase/supabase-js version 1.21.0? Using that version should work, what code exactly are you trying?

**Reply by 72L** [reply]
I've upgraded to the latest version (1.22.6) and this is the response:

```
{
  data: null,
  error: {
    statusCode: '42501',
    error: '',
    message: 'new row violates row-level security policy for table "objects"'
  }
}
```

I don't have any row-level security policy nor do I have a table called "objects" <- assume this is an internal table. What do I need to do?

**Reply by thebengeu** [reply]
You need to [add a security rule](https://supabase.io/docs/guides/storage#add-security-rules) by going to Storage -> Policies to allow INSERT in order for upload with an anonymous user to work, e.g. with an SQL conditional expression of:

```
bucket_id = 'my-bucket'
AND role() = 'anon'
```

![Screenshot 2021-09-23 at 11 45 55](https://user-images.githubusercontent.com/1154867/134452097-c7cf0c2d-e538-488b-914f-660018a1e971.png)
